core/src/main/java/org/infinispan/persistence/sifs/Index.java

```diff
@@ -245,6 +286,41 @@ public void deleteFileAsync(int fileId) {
    }
 
    public CompletionStage<Void> stop() throws InterruptedException {
+
+      CompletionStage<Void> statsStage = CompletableFuture.runAsync(() -> {
+         ConcurrentMap<Integer, Compactor.Stats> map = compactor.getFileStats();
+         File statsFile = new File(indexDir.toFile(), "index.stats");
+         try {
+            FileChannel statsChannel = new RandomAccessFile(statsFile, "rw").getChannel();
+            statsChannel.truncate(0);
+            // Maximum size that all ints and long can add up to
+            ByteBuffer buffer = ByteBuffer.allocate(4 + 4 + 4 + 8);
+            for (Map.Entry<Integer, Compactor.Stats> entry : map.entrySet()) {
+               int file = entry.getKey();
+               int total = entry.getValue().getTotal();
+               if (total == -1) {
+                  total = (int) fileProvider.getFileSize(file);
+               }
+               int free = entry.getValue().getFree();
+               if (total == free) {
+                  log.tracef("Deleting file %s since it has no free bytes in it", file);
+                  // No reason to keep an empty file around
+                  fileProvider.deleteFile(file);
+                  continue;
+               }
+               buffer.putInt(file);
+               buffer.putInt(total);
+               buffer.putInt(free);
+               buffer.putLong(entry.getValue().getNextExpirationTime());
+               buffer.flip();
+               write(statsChannel, buffer);
+               buffer.flip();
+            }
+         } catch (IOException e) {
+            throw new CacheException(e);
+         }
+      }, blockingExecutor);
+
       for (FlowableProcessor<IndexRequest> flowableProcessor : flowableProcessors) {
```

Comment:
don't you want to complete the processor before storing the stats?

Reply:
Yeah, actually it should be part of the `thenRun` below.